internal/batch/batch_processor.go

```diff
@@ -568,11 +601,53 @@ func (bp *batchProcessor) dispatchBatch(payload *DispatchPayload) error {
 	// Call the dispatcher to do the heavy lifting - will only exit if we're closed
 	return operations.RunWithOperationContext(bp.ctx, func(ctx context.Context) error {
 		return bp.retry.Do(ctx, "batch dispatch", func(attempt int) (retry bool, err error) {
-			return true, bp.conf.dispatch(ctx, payload)
+			err = bp.conf.dispatch(ctx, payload)
+			if err != nil {
+				if bp.isCancelled() {
+					log.L(ctx).Warnf("Batch %s was cancelled - replacing with gap fill", payload.Batch.ID)
+					for _, msg := range payload.Messages {
+						if err := bp.writeGapFill(ctx, msg); err != nil {
+							return true, err
+						}
+					}
+					err = nil
+					payload.State = core.MessageStateCancelled
+				}
+			} else {
+				if core.IsPinned(bp.conf.txType) {
+					payload.State = core.MessageStateSent
+				} else {
+					payload.State = core.MessageStateConfirmed
+				}
+			}
+			return true, err
 		})
 	})
 }
 
+func (bp *batchProcessor) writeGapFill(ctx context.Context, msg *core.Message) error {
+	// Gap fill is only needed for private messages
+	if msg.Header.Type != core.MessageTypePrivate {
```

Comment:
I went back and forth on whether to send gap fill messages for broadcasts. Ultimately they are needed for private messages for two reasons:
1. This sender has "spent" a nonce, and may have calculated and used nonces after it. All recipients need to know about the spent nonce so they can process later messages and get back in sync.
2. All recipients will have received the message payload via data exchange, and will have created a message in "pending" state. They need to know to move those messages to "cancelled" so that they don't stay pending forever.

Neither of these is true for broadcasts. Therefore it felt like cancelling a broadcast can be a "local only" operation, and broadcasting a gap fill to everyone might be more confusing.